### PR TITLE
Fix OOB read in max_pool2d/max_pool3d with overflow-scale dilation

### LINKED ISSUE #189664: [CUDA] illegal memory access on max_pool2d/max_pool3d with integer-overflow dilation
### 🐛 Describe the bug

The OOB is caused by dilation=INT32_MAX overflowing the 32-bit loop counter in the CUDA max-pooling kernel.
```python
# Bug: max_pool_forward_nchw 32-bit index overflow on huge dilation
import torch
print("torch:", torch.__version__)
torch.cuda.init()
x = torch.zeros(5, 10, 24, 24, device='cuda')
torch.nn.functional.max_pool2d(x, kernel_size=1, stride=None, padding=0,
                               dilation=2147483647, ceil_mode=True)
torch.cuda.synchronize()
print("done (no error)")
```

```text
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void at::native::<unnamed>::max_pool_forward_nchw<float, int>(T2, const T1 *, long, long, long, int, int, int, int, int, int, int, int, int, int, T1 *, long *)+0xca0
=========     by thread (1,0,0) in block (0,0,0)
=========     Address 0x7fd6c5e00000 is out of bounds
=========     and is 8589934592 bytes before the nearest allocation at 0x7fd8c5e00000 of size 2097152 bytes
```

### Versions

Collecting environment information...

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @malfet @mikaylagawarecki

### PR DESCRIPTION
**Pull request by MDSALMANSHAMS**
Fixes #189664.

`max_pool2d`/`max_pool3d` with an overflow-scale `dilation` reads out of bounds on CUDA. This is now
fixed inside the CUDA kernels, so no caller loses an input that works today.

**This replaces the earlier version of this PR**, which added an `inputSize - 1 + dilation <= INT32_MAX`
check to the shared `pool2d/3d_shape_check` (C++ and meta). That guard was wider than the bug: at
`dilation == 1` it collapses to a bound on input size alone, which also hits `avg_pool2d/3d` and the
int64-clean CPU `max_pool` kernels, where there is no overflow to prevent — a real BC break, reachable
on meta under `torch.compile`. `aten/src/ATen/native/Pool.h` and `torch/_meta_registrations.py` are
back to their unmodified state; nothing in this PR touches the shape checks or the meta layer.

## Root cause

Not the effective window extent. The repro uses `kernel_size=1`, where the extent
`dilation * (kernel_size - 1) + 1` is just `1` and cannot overflow.

The defect is in the window traversal. `max_pool_forward_nchw`
(`aten/src/ATen/native/cuda/DilatedMaxPool2d.cu`) declared its loop counters as plain `int` while the
bounds around them are `index_t`:

```cpp
for (int h = hstart; h < hend; h += dilation_h) {
  for (int w = wstart; w < wend; w += dilation_w) {
    scalar_t val = btm_data[h * width + w];
```

`h` starts inside the input and advances by `dilation_h`, so `h += dilation_h` signed-overflows and
wraps negative. `h < hend` then stays true and `btm_data[h * width + w]` reads roughly 2^31 elements
before the allocation, matching the sanitizer output in the issue exactly (`8589934592` bytes = 2^31 x 4).

Overflow begins at `inputSize - 1 + dilation > INT_MAX`. Measured on an A40:

| input H | last dilation that works | first dilation that faults |
| --- | --- | --- |
| 24 | 2147483624 | 2147483625 |
| 100 | 2147483548 | 2147483549 |
| 500 | 2147483148 | 2147483149 |

which is `INT_MAX - (H - 1)` in each case.

### max_pool2d — widen the counters, and let the existing `index_t` dispatch do the rest

This is the narrower fix suggested in review, and it is what the kernel family already supports:
`max_pool_forward_nchw` / `max_pool_forward_nhwc` are templated on `index_t` and **both `int32_t` and
`int64_t` instantiations are already emitted**, selected by `can_use_int32_nhwc` / `can_use_int32_nchw`.

- `max_pool_forward_nchw`: the two loop counters `int h` / `int w` become `index_t` (the reported OOB).
- `can_use_int32_nhwc` / `can_use_int32_nchw` take `dilation_h` / `dilation_w` and return `false` when a
  window position would not fit in `int32_t`, so an overflowing dilation simply selects the `int64_t`
  instantiation that is already compiled.

No new instantiations, no new user-facing restriction: the op now returns the correct result where it
used to read OOB. `max_pool_backward_nchw` was already fully `index_t` and needed no change.

While widening the eligibility test I found a **second overflow the review did not cover**: the prefetch
path in `max_pool_forward_nhwc` computes `ih*dilation_h` as `int * int` before widening, over a fixed
`MAXh x MAXw` unrolled window, so positions reach `2 * dilation` rather than one step. That path is
`#if defined(USE_ROCM)`, so it is only exercised by ROCm CI. Those products are now formed in `index_t`,
and the eligibility bound uses `max_window_steps = 2` to cover both kernels.

### max_pool3d — reject at the launcher, CUDA-only

The 3d kernels are a different situation: they do all window indexing in 32-bit arithmetic and take the
input extents as plain `int`, with **no `int64_t` instantiation to fall back to** (see the note on #52822).
Widening them is a much larger change than this bug fix. So the CUDA launcher
(`max_pool3d_with_indices_out_cuda_template`) rejects an input whose dilation would overflow that
indexing, with an explicit message.

Deliberately **not** placed in the shared `pool3d_shape_check`: that would put the restriction back on CPU
and on `avg_pool3d`, which is the problem this rework exists to avoid. Only the CUDA path is constrained,
and only where the kernel genuinely cannot represent the index.

Only the trailing advance needs checking — the window extent `(kernel - 1) * dilation` is already bounded
by the input size, since a larger extent makes the output size non-positive and `pool3d_shape_check`
rejects it first.

## Tests

- `test_max_pool2d_large_dilation_no_overflow` (CUDA): `dilation = INT_MAX` in each of three
  configurations, contiguous and channels-last, must return the correct result — not raise.
- `test_max_pool3d_large_dilation_rejected` (CUDA): overflow-scale dilation raises, **and the largest
  dilation that still fits is asserted to succeed**, so a later off-by-one tightening cannot silently
  narrow the supported range (review item 6).
- `test_pool_large_spatial_dim_undilated`: `avg_pool2d/3d` and `max_pool2d/3d` with a spatial dim of
  `2**31 + 1` and `dilation = 1` still work, checked on `device="meta"` so nothing is allocated. This is a
  regression test for the BC break of the previous revision — it fails against it and passes here.
- `error_inputs_max_pool3d` gains the CUDA-only overflow case (`torch.device(device).type`); the
  `max_pool2d` `error_inputs` from the previous revision is **removed**, because 2d no longer errors.

Scoping the claim made in the previous description: this is not "every backend rejects with an identical
message". `max_pool2d` short-circuits to `at::quantized_max_pool2d` / `at::mkldnn_max_pool2d`
(`native/Pooling.cpp`), neither of which reaches these kernels; quantized validates dilation separately and
mkldnn rejects `dilation != 1`. The fix is scoped to the CUDA kernels that had the defect.

## Verification (A40, CUDA 12.4, built from source)

- `compute-sanitizer --tool memcheck` on the issue's repro: **0 errors** (previously an OOB read of
  8589934592 bytes), and the result is now correct.
- `test/nn/test_pooling.py`: **207 passed, 59 skipped**, including the three tests above.
  (The `*_large_size_*` tests that allocate multi-GB tensors were excluded on this box for memory, not
  because of this change.)
- `test_ops.py -k "test_errors and max_pool"`: **12 passed**, CPU and CUDA.
- Both `.cu` files compile with zero new warnings.

### CODE REVIEW

**Comment by MDSALMANSHAMS** [comment]
@malfet this is the PR for #189664 - the check went in at the meta layer as you suggested, plus the shared shape-check header so the eager path and the OpInfo `error_inputs` agree with it.

It's stuck before CI though: all 5 workflow runs on `ab9b134f` are `action_required`, so there's exactly 1 check-run on the head and nothing has actually run. Could you approve the pending workflows when you get a chance?

Flagging why I care about the order here: #188328 landed on a stale green last week for exactly this reason - its workflows sat un-approved for weeks, the inductor shards never ran on it, and it got auto-reverted 8 hours after merging. I'd rather this one run properly first.

Files: `aten/src/ATen/native/Pool.h` (+19), `torch/_meta_registrations.py` (+30), `torch/testing/_internal/common_methods_invocations.py` (+20).

**Comment by MDSALMANSHAMS** [comment]
@albanD @mruberry @malfet — this one needs a maintainer to approve the pending workflow runs. It's a one-click unblock; nothing here is failing.

State: the 5 workflow runs on `ab9b134f` (`pull`, `Lint`, `BC Lint`, `docs-build`, `Auto Request Review`) have been sitting at `action_required` since Aug 3, so CI has never actually run. The only check-run on the head is `Meta Internal-Only Changes Check` (success), which makes the combined status look green when it is really un-run.

The change is scoped to exactly what was asked for on #189664:

- validation at the meta layer (`torch/_meta_registrations.py`), so eager and compile agree on the error — per @malfet's Aug 2 note that a meta-layer check is sufficient
- the guard in the shared shape checks in `aten/src/ATen/native/Pool.h`, so CPU and CUDA report the same error rather than only CUDA crashing
- `error_inputs` on the `max_pool2d` / `max_pool3d` OpInfos

3 files, +69/-0. No rebase needed — `Pool.h` has had no upstream commits since the base commit and the PR is still mergeable. Happy to run `@pytorchbot merge` myself once CI is green.

**Comment by MDSALMANSHAMS** [comment]
@albanD @malfet @mruberry — following up on this one: it still has no CI results at all.

The combined status reads green, but that comes from a single check-run (`Meta Internal-Only Changes Check`). The actual workflows have been sitting in the "approve and run workflows" state since 2026-08-04, so the test suite has never executed on this branch.

Could one of you approve the pending run? The change fixes #189664 (`actionable`): a device-agnostic guard in `aten/src/ATen/native/Pool.h` plus `error_inputs` coverage, with no host-to-device sync introduced. I'd like real CI signal on it before taking any more of your review time.

**Comment by malfet** [comment]
@claude review this

**Comment by MDSALMANSHAMS** [comment]
Thanks for the review — the blocker was right, and I've reworked the PR rather than patched the guard.

The shared-shape-check approach is gone entirely: `Pool.h` and `_meta_registrations.py` are back to
unmodified, so there is no new bound on input size, nothing changes for `avg_pool2d/3d` or the CPU
`max_pool` kernels, and the meta layer is untouched. The overflow is now fixed where it happens.

- **2d**: the `index_t` route you pointed at. `max_pool_forward_nchw`'s two plain-`int` loop counters
  become `index_t`, and `can_use_int32_nchw`/`can_use_int32_nhwc` now take the dilations so an
  overflowing window position selects the `int64_t` instantiation that is already emitted. No new
  instantiations, no user-facing restriction — the op returns the correct result instead of raising.
- **3d**: those kernels index in 32-bit and take extents as plain `int` with no int64 fallback
  (#52822), so there is nothing to widen to. The CUDA launcher rejects the input instead —
  deliberately *not* in `pool3d_shape_check`, so CPU and `avg_pool3d` are unaffected.

Also picked up while widening the eligibility test: the `USE_ROCM` prefetch path in
`max_pool_forward_nhwc` forms `ih*dilation_h` as `int * int` before widening, over a fixed `MAXh x MAXw`
window, so it reaches two dilation steps rather than one. That is now formed in `index_t` and the
eligibility bound accounts for it.

On the smaller points: added a test that the largest still-representable dilation is accepted (6), the
2d `error_inputs` entry is removed since 2d no longer errors, and the "every backend" claim in the
description is scoped — quantized/mkldnn short-circuit before these kernels (5). There is also a new
regression test that `avg_pool2d/3d` and `max_pool2d/3d` accept a spatial dim above 2^31 with
`dilation=1`, on meta so nothing is allocated; it fails against the previous revision of this PR.

Verified locally on an A40 build: `compute-sanitizer memcheck` on the issue repro is clean (0 errors,
was an 8589934592-byte OOB read), `test/nn/test_pooling.py` is 207 passed / 59 skipped, and
`test_ops.py -k "test_errors and max_pool"` is 12 passed on CPU and CUDA. CI still needs a workflow
approval to run.

**Comment by MDSALMANSHAMS** [comment]
@malfet — the rework addressing your bot review's blocker (dropped the shared shape-check approach entirely) went up on 08-18. Workflows on this head (`4fa997cf`) are still `action_required` and CI hasn't run since. Whenever you get a chance to approve the pending run, happy to address anything it turns up.

**Comment by MDSALMANSHAMS** [comment]
@malfet — checking back in on the 08-18 rework, and a question given #195547: this PR is a single-operator `TORCH_CHECK` guard (max_pool2d/3d dilation overflow), which is exactly the shape you flagged there as likely to be declined in favor of a device-agnostic, schema-derived fix. Does this one still stand as-is, or would you rather it fold into the structured-kernel migration direction? Happy to adjust either way — also noting CI has still never run on this head (`4fa997cf`), only the Meta-internal check, so it's still behind the workflow-approval gate too.
